src/sagemaker_xgboost_container/algorithm_mode/dask_training.py

```diff
@@ -0,0 +1,174 @@
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import logging
+import os
+import socket
+from subprocess import Popen
+import sys
+import time
+
+import dask.dataframe as dask_dataframe
+from dask.dataframe import DataFrame
+from dask.dataframe import Series
+from dask.distributed import Client, wait
+from sagemaker_algorithm_toolkit import exceptions as exc
+import xgboost as xgb
+
+from sagemaker_xgboost_container.constants.xgb_constants import MODEL_NAME
+
+DASK_PATH = "/miniconda3/bin"
+SCHEDULER_PORT = "8786"
+
+WAIT_TIME_DAEMONS_STARTUP_SEC = 2
+WAIT_FOR_ALL_WORKERS_TIMEOUT_SEC = 20
+
+logger = logging.getLogger(__name__)
+
+
+def _start_daemons(scheduler_host_ip: str, current_host_ip: str):
+    cmd_start_scheduler = os.path.join(DASK_PATH, "dask-scheduler")
+    cmd_start_worker = os.path.join(DASK_PATH, "dask-cuda-worker")
+    schedule_conn_string = "tcp://{}:{}".format(scheduler_host_ip, SCHEDULER_PORT)
+
+    # We can set this based on number of vCPUs in the future.
+    # So far haven't seen improvement with > 4, during testing.
+    num_threads_per_worker_process = "4"
+    # Dask cuda worker API doc: https://docs.rapids.ai/api/dask-cuda/nightly/api.html
+    worker_cli_command = [cmd_start_worker, schedule_conn_string, "--no-dashboard",
+                          "--nthreads", num_threads_per_worker_process]
+    # Dask distributed scheduler API doc: https://docs.dask.org/en/stable/deploying-cli.html
+    scheduler_cli_command = [cmd_start_scheduler, "--no-dashboard"]
+    if scheduler_host_ip == current_host_ip:
+        Popen(scheduler_cli_command)
+    Popen(worker_cli_command)
+
+
+def _get_host_ip(host_name: str) -> str:
+    ip_wait_time = 5
+    counter = 0
+    host_ip = ""
+    from socket import gaierror
+    while counter < ip_wait_time and host_ip == "":
+        try:
+            host_ip = socket.gethostbyname(host_name)
+            break
+        except gaierror:
+            counter += 1
+            time.sleep(1)
+
+    if counter == ip_wait_time and host_ip == "":
+        raise Exception(
+            "Exceeded max wait time of {}s for hostname resolution for host {}.".format(ip_wait_time, host_name)
+        )
+    return host_ip
+
+
+def _read_data(client: Client, local_path: str, content_type: str) -> (DataFrame, Series):
+    if content_type == "csv":
+        dataframe = dask_dataframe.read_csv(local_path + "/*.csv", header=None)
+    # content_type should only be csv or parquet after checks in train.py.
+    else:
+        dataframe = dask_dataframe.read_parquet(local_path)
+    target_column = dataframe.columns[0]
+    labels = dataframe[target_column]
+    features = dataframe[dataframe.columns.difference([target_column])]
+
+    features, labels = client.persist(
+        [features, labels]
+    )
+    wait([features, labels])
+
+    return features, labels
+
+
+def run_training_with_dask(
```

Comment:
Within this module, would it be better to group things together under a class or separate more common utilities out into a separate module where reusability may apply?

Within this function, would it be possible to separate out data loading and execution?

Reply:
Did some refactoring.